### PR TITLE
Test corner cases for all operators

### PR DESCRIPTION
**Pull request by hugomg**

### CODE REVIEW

File: test_operators.lua [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            local ok, err = check(f_lua, f_titan, {x, y})
+            if not ok then
+                error(string.format("%s %s %s: %s",
+                    tostring(x), op_str, tostring(y), err))
+            end
+        end
+    end
+end
+
+return {
```
Comment (gligneul) [inline test_operators.lua]: I think we should use the protocol we are already using in other modules:
```
local module = {}

function module.foo()
end

return module

File: test_operators.lua [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+end
+
+local function are_same(a, b)
+    return (a == b) or (isnan(a) and isnan(b))
+end
+
+local function check(f_lua, f_titan, args)
+    local ok1, a = pcall(f_lua, table.unpack(args))
+    local ok2, b = pcall(f_titan, table.unpack(args))
+    if ok1 ~= ok2 then
```
Comment (gligneul) [inline test_operators.lua]: If both calls crash this will be false and the values will be compared bellow, is this expected?
Shouldn't we do check something like:
```
if not ok1 or not ok2 then
```

File: test_operators.lua [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+local function isnan(x)
+    return x ~= x
+end
+
+local function are_same(a, b)
+    return (a == b) or (isnan(a) and isnan(b))
+end
+
+local function check(f_lua, f_titan, args)
+    local ok1, a = pcall(f_lua, table.unpack(args))
```
Comment (gligneul) [inline test_operators.lua]: ok1 and ok2 are not good variable names, we could be a bit more specific like ok_lua and ok_titan.

File: test_operators.lua [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            (ok2 and "didn't crash" or "crashed"))
+    end
+    if ok1 and ok2 and not are_same(a, b) then
+        return false, string.format("(lua: %s, titan: %s)",
+            tostring(a), tostring(b))
+    end
+    return true
+end
+
+local function check_unop(op_str, f_lua, f_titan, typ1)
```
Comment (gligneul) [inline test_operators.lua]: Why typ1 instead of typ? Seems a typo but I'm not sure.

File: test_operators.lua [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    for _, x in ipairs(values[typ1]) do
+        local ok, err = check(f_lua, f_titan, {x})
+        if not ok then
+            error(string.format("%s %s: %s",
+                op_str, tostring(x), err))
+        end
+    end
+end
+
+local function check_binop(op_str, f_lua, f_titan, typ1, typ2)
```
Comment (gligneul) [inline test_operators.lua]: Again, we could replace 1 and 2 with better names. Suggestion:
```
s/typ1/lhs_type/
s/typ2/rhs_type/
```

File: test_operators.lua [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+local function isnan(x)
+    return x ~= x
+end
+
+local function are_same(a, b)
+    return (a == b) or (isnan(a) and isnan(b))
+end
+
+local function check(f_lua, f_titan, args)
```
Comment (gligneul) [inline test_operators.lua]: Why manually implement vararg instead of using `...`?

File: test_operators.lua (hunk 3 above)
Comment (gligneul) [reply test_operators.lua]: `a` and `b` are also bad names, why not `ret_lua` (or `*_lua`)?

File: test_operators.lua [hunk 7]
```diff
@@ -0,0 +1,61 @@
+-- Used in checker_spec
+
+local values = {
+    ["boolean"] = { true, false },
+    ["integer"] = {
+        math.mininteger, -10, -3, -2, -1, 0, 1, 2, 3, 10, math.mininteger,
```
Comment (gligneul) [inline test_operators.lua]: Is testing -3 and 3 really useful?

File: test_operators.lua [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    },
+    ["float"] = {
+        -math.huge, -2.0, -1.0, 0.0, 1.0, math.pi
```
Comment (gligneul) [inline test_operators.lua]: We could test +math.huge also?

File: spec/coder_spec.lua [hunk 9]
```diff
@@ -284,7 +284,7 @@ describe("Titan coder /", function()
             end
         ]]))
 
-        it("local variables", function()
+               it("local variables", function()
```
Comment (gligneul) [inline spec/coder_spec.lua]: Indentation is wrong here.

File: spec/coder_spec.lua [hunk 10]
```diff
@@ -298,315 +298,126 @@ describe("Titan coder /", function()
...
-                run_binop_test("lshift", "<<", "0xdead", "-1")
-                run_binop_test("lshift", "<<", "0xdead", " 100")
-                run_binop_test("lshift", "<<", "0xdead", "-100")
-                run_binop_test("lshift", "<<", "0xdead", "math.maxinteger")
-                run_binop_test("lshift", "<<", "0xdead", "math.mininteger")
-            end)
-        end)
+        setup_unop("neg_int", "-",   "integer", "integer")
+        setup_unop("bnot",    "~",   "integer", "integer")
+        setup_unop("not_bool","not", "boolean", "boolean")
```
Comment (gligneul) [inline spec/coder_spec.lua]: Missing a space after the first comma.

File: spec/coder_spec.lua [hunk 11]
```diff
@@ -298,315 +298,126 @@ describe("Titan coder /", function()
...
+
+        setup_binop("add_int"       , "+", "integer", "integer", "integer")
```
Comment (gligneul) [inline spec/coder_spec.lua]: In the previous block the comma was aligned with argument in the left, here it is aligned in the right. We should be consistent.
Also, we could align the other arguments.

**Review by gligneul — CHANGES_REQUESTED** [review]

**Comment by gligneul** [comment]
Github is showing some of my comments as outdated because I reviewed them commit by commit, please take a look in all of them.

**Review by gligneul — APPROVED** [review]
